compiler/rustc_parse/src/parser/expr.rs

```diff
@@ -2466,11 +2466,26 @@ impl<'a> Parser<'a> {
         if !self.eat_keyword(kw::In) {
             self.error_missing_in_for_loop();
         }
+        let expr_span = self.token.span;
         self.check_for_for_in_in_typo(self.prev_token.span);
         let expr = self.parse_expr_res(Restrictions::NO_STRUCT_LITERAL, None)?;
 
         let pat = self.recover_parens_around_for_head(pat, begin_paren);
 
+        // Recover from missing expression in `for` loop
+        if matches!(expr.kind, ExprKind::Block(..))
+            && !matches!(self.token.kind, token::OpenDelim(token::Delimiter::Brace))
+            && self.may_recover()
+        {
+            let mut diag = self.sess.struct_err("expected expression, found block");
+            diag.set_span(expr_span);
+            diag.help(
+                "try adding an expression to the `for` loop: `for PATTERN in EXPRESSION { BODY }`",
+            );
+            self.mk_expr(expr_span, ExprKind::Err);
+            return Err(diag);
```

Comment:
Copying this from the resolved comment:

> We should be able to recover by returning a for loop that looks like for pat in <ERROR> { expr }, where we reinterpret the expr we parsed on line 2471 as a block, and insert an error expression instead. You can make an error expression with this.mk_expr(span, ExprKind::Err)).

When I said that, I meant -- instead of returning `Err`, we can *emit* the diagnostic, and instead return `Ok(self.mk_expr(...))`, with a properly constructed `for` loop expression. We would use the `ExprKind::Err` expression in the `EXPRESSION` part of `for PATTERN in EXPRESSION { BODY }`.